lib/src/rules/tighten_type_of_initializing_formals.dart

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2020, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/analysis/features.dart';
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/token.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+
+import '../analyzer.dart';
+
+const _desc = r'Tighten type of initializing formal.';
+
+const _details = r'''
+
+Tighten type of initializing formal if an non null assert exists.
+
+**BAD:**
+```
+class A {
+  A.c1(this.p) : assert(p != null);
+  A.c2(this.p);
+  final String? p;
+}
+```
+
+**GOOD:**
+```
+class A {
+  A.c1(String this.p) : assert(p != null);
+  A.c2(this.p);
+  final String? p;
+}
+```
+
+''';
+
+class TightenTypeOfInitializingFormals extends LintRule
+    implements NodeLintRule {
+  TightenTypeOfInitializingFormals()
+      : super(
+          name: 'tighten_type_of_initializing_formals',
+          description: _desc,
+          details: _details,
+          group: Group.style,
+        );
+
+  @override
+  void registerNodeProcessors(NodeLintRegistry registry,
+      [LinterContext context]) {
+    final visitor = _Visitor(this, context);
+    registry.addCompilationUnit(this, visitor);
+    registry.addConstructorDeclaration(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  _Visitor(this.rule, this.context);
+
+  final LintRule rule;
+  final LinterContext context;
+
+  bool _isNonNullableEnabled;
+
+  @override
+  void visitCompilationUnit(CompilationUnit node) {
+    _isNonNullableEnabled = node.featureSet.isEnabled(Feature.non_nullable);
+  }
+
+  @override
+  void visitConstructorDeclaration(ConstructorDeclaration node) {
+    if (!_isNonNullableEnabled) return;
+
+    node.initializers
+        .whereType<AssertInitializer>()
+        .map((e) => e.condition)
+        .whereType<BinaryExpression>()
+        .where((element) => element.operator.type == TokenType.BANG_EQ)
+        .map((e) => e.rightOperand is NullLiteral
+            ? e.leftOperand
+            : e.leftOperand is NullLiteral ? e.rightOperand : null)
+        .where((e) => context.typeSystem.isNullable(e.staticType))
```

Comment:
As far as I can tell, `e` can be `null`, which will throw in `.staticType`.

I think this will happen when neither operand is `null`:
```
C(this.p, this.q): assert(p != q);
```

Reply:
Fixed.